src/java_bytecode/java_static_initializers.cpp

```diff
@@ -0,0 +1,186 @@
+/*******************************************************************\
+
+Module: Java Static Initializers
+
+Author: Chris Smowton, chris.smowton@diffblue.com
+
+\*******************************************************************/
+
+#include "java_static_initializers.h"
+#include <goto-programs/class_hierarchy.h>
+#include <util/std_types.h>
+#include <util/std_expr.h>
+#include <util/std_code.h>
+
+#include <iostream>
+
+/// Get the Java static initializer wrapper name for a given class (the wrapper
+/// checks if static initialization has already been done before invoking the
+/// real static initializer if not).
+/// Doesn't check whether the symbol actually exists
+/// \param class_name: class symbol name
+/// \return static initializer wrapper name
+irep_idt clinit_wrapper_name(const irep_idt &class_name)
+{
+  return id2string(class_name) + "::clinit_wrapper";
+}
+
+/// Get name of the static-initialization-already-done global variable for a
+/// given class.
+/// \param class_name: class symbol name
+/// \return static initializer wrapper-already run global name
+static irep_idt clinit_already_run_variable_name(const irep_idt &class_name)
+{
+  return id2string(class_name) + "::clinit_already_run";
+}
+
+/// Get name of the real static initializer for a given class. Doesn't check
+/// if a static initializer actually exists.
+/// \param class_name: class symbol name
+/// \return Static initializer symbol name
+static irep_idt clinit_function_name(const irep_idt &class_name)
+{
+  return id2string(class_name) + ".<clinit>:()V";
+}
+
+/// Checks whether a static initializer wrapper is needed for a given class,
+/// i.e. if the given class or any superclass has a static initializer.
+/// \param class_name: class symbol name
+/// \param symbol_table: global symbol table
+/// \return true if a static initializer wrapper is needed
+static bool needs_clinit_wrapper(
+  const irep_idt &class_name, const symbol_tablet &symbol_table)
+{
+  if(symbol_table.has_symbol(clinit_function_name(class_name)))
+    return true;
+
+  const class_typet &class_type =
+    to_class_type(symbol_table.lookup_ref(class_name).type);
+
+  for(const class_typet::baset &base : class_type.bases())
+  {
+    if(symbol_table.has_symbol(
+         clinit_wrapper_name(to_symbol_type(base.type()).get_identifier())))
+    {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+/// Creates a static initializer wrapper for the given class. This function will
+/// call static initializer wrappers for all parent classes that have them, then
+/// run this class' clinit method if one exists.
+/// \param class_name: class symbol name
+/// \param symbol_table: global symbol table; will gain a clinit_wrapper symbol
+///   and a corresponding has-run-already global.
+static void create_clinit_wrapper(
+  const irep_idt &class_name, symbol_tablet &symbol_table)
+{
+  // Otherwise, assume that class C extends class C' and implements interfaces
```

Comment:
Otherwise to what? If this follows on from the function comment it should probably be part of it